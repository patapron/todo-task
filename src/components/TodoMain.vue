<template>
  <ul class="todo-list">
    <li
      :class="{ completed: obj.completed }"
      v-for="(obj, index) in list"
      :key="obj.id"
    >
      <div class="view">
        <input class="toggle" type="checkbox" v-model="obj.completed" />
        <label>{{ obj.name }}</label>
        <button class="destroy" @click="deleteTask(obj.id)"></button>
      </div>
    </li>
  </ul>
</template>

<script>
export default {
  props: ["list"],
  methods: {
    /**
     * Delete task by id emit
     */
    deleteTask(taskId) {
      this.$emit("deleteTaskEmit", taskId);
    },
  },
};
</script>
