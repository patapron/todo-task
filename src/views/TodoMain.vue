<template>
  <ul class="todo-list">
    <li
      :class="{ completed: obj.completed, editing: isEditing(obj) }"
      v-for="(obj, index) in list"
      :key="obj.id"
    >
      <div class="view">
        <input
          :disable="editing"
          class="toggle"
          type="checkbox"
          v-model="obj.completed"
        />
        <label
          v-if="!editing || (editing && !isEditing(obj))"
          :disable="editing"
          @dblclick.stop="editTask(obj)"
          >{{ obj.name }}</label
        >
        <button class="destroy" @click="deleteTask(obj.id)"></button>
      </div>
      <input
        class="edit"
        type="text"
        v-model="obj.name"
        v-on:keyup.enter="updateName()"
        focused:from="isEditing(obj)"
        @blur="cancelEdit()"
      />
    </li>
  </ul>
</template>
<script>


export default {
  name: "TodoMain",
  props: {
    list: [],
  },
  data() {
    return {
      editing: any,
      backupName: '',
    };
  },
  methods: {
    /**
     * Check if this task arr editing
     */
    isEditing:function(task) {
      return this.editing !== null && task === this.editing;
    },
    /**
     * Reset editing task name
     */
    cancelEdit() {
      console.log("cancel");
      if (this.editing) {
        this.editing.name = this.backupName;
      }
      this.backupName = "";
      this.editing = null;
    },
    /**
     * Update task name emiter
     */
    updateName() {
      this.$emit("changeTaskEmit", this.editing);
      this.backupName = "";
      this.editing = null;
    },
    /**
     * Set edit object and old copy value
     */
    editTask(task) {
      this.backupName = task.name;
      this.editing = task;
    },
    /**
     * Delete task by id emit
     */
    deleteTask(taskId) {
      this.$emit("deleteTaskEmit", taskId);
    },
  },
};
</script>
