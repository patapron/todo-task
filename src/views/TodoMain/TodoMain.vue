<template>
  <ul class="todo-list">
    <li
      :class="{ completed: obj.completed, editing: isEditing(obj) }"
      v-for="obj in list"
      :key="obj.id"
    >
      <div class="view">
        <input
          :disable="editing"
          class="toggle"
          type="checkbox"
          v-model="obj.completed"
        />
        <label
          v-if="!editing || (editing && !isEditing(obj))"
          :disable="editing"
          @dblclick.stop="editTask(obj)"
          >{{ obj.name }}</label
        >
        <button class="destroy" @click="deleteTask(obj.id)"></button>
      </div>
      <input
        class="edit"
        type="text"
        v-model="obj.name"
        v-on:keyup.enter="updateName()"
        focused:from="isEditing(obj)"
        @blur="updateName()"
        @keydown.esc="cancelEdit()"
      />
    </li>
  </ul>
</template>
<script lang="ts">
import "./TodoMain.css";
import { Task } from "@/typings/types";
import Vue, { PropType } from "vue";

export default Vue.extend({
  props: {
    list: {
      type: [] as PropType<Task[] | Function>,
    },
  },
  data() {
    return {
      // editing: Task as {},
      editing: null as Task | null,
      backupName: "",
    };
  },
  methods: {
    /**
     * Check if this task arr editing
     */
    isEditing: function (task: any): any {
      return this.editing !== null && task === this.editing;
    },
    /**
     * Reset editing task name
     */
    cancelEdit() {
      if (this.editing) {
        this.editing.name = this.backupName;
      }
      this.backupName = "";
      this.editing = null;
    },
    /**
     * Update task name emiter
     */
    updateName() {
      if(!this.editing?.name){
        this.$emit("deleteTaskEmit", this.editing?.id);
      }else{
        this.$emit("changeTaskEmit", this.editing);
      }
      this.backupName = "";
      this.editing = null;
    },
    /**
     * Set edit object and old copy value
     */
    editTask(task: Task) {
      this.backupName = task.name;
      this.editing = task;
    },
    /**
     * Delete task by id emit
     */
    deleteTask(taskId: number) {
      this.$emit("deleteTaskEmit", taskId);
    },
  },
});
</script>
