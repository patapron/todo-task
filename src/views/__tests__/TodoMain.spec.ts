import TodoMain from "../TodoMain.vue";
import { shallowMount } from "@vue/test-utils";

describe("TodoMain", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(TodoMain, {
      propsData: {
        list: [],
      },
      mocks: {
        $t: () => "Test text",
        $can: () => true,
      },
    });
  });

  test("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
