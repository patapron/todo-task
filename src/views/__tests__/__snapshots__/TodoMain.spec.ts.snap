// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoMain renders correctly 1`] = `<ul class="todo-list"></ul>`;
